[1822] First pass at the first stockround

magnusfalt opened the pull request:
First pass at the stock round which implementes a new type of auction, bidbox.

crericha-qualitrol commented on assets/app/view/game/player.rb:
```diff
@@ -101,6 +101,13 @@ def render_info
+          trs.concat([
+             h(:tr, [
+               h(:td, 'Bidding tokens'),
+               h('td.right', "#{@game.active_step.bidding_tokens(@player)} / #{@game.bidding_token_per_player}"),
```
Using round state for bidding_tokens would be better than active_step.

crericha replied:
Using round state for bidding_tokens would be better than active_step.

crericha-qualitrol commented on assets/app/view/game/player.rb:
```diff
@@ -101,6 +101,13 @@ def render_info
+               h(:td, 'Bidding tokens'),
+               h('td.right', "#{@game.active_step.bidding_tokens(@player)} / #{@game.bidding_token_per_player}"),
+             ]),
+           ]) if @game.active_step&.current_actions&.include?('bidding_tokens')
```
Suggest using the bid action and gate on respond_to?(:bidding_tokens)

crericha replied:
Suggest using the bid action and gate on respond_to?(:bidding_tokens)

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -57,6 +57,12 @@ def render
+          children << render_bidbox(@game.bidbox_concessions) if @current_actions.include?('bidbox_concessions') &&
+            @game.respond_to?(:bidbox_concessions)
+          children << render_bidbox(@game.bidbox_privates) if @current_actions.include?('bidbox_privates') &&
+            @game.respond_to?(:bidbox_privates)
```
Why do these need to be actions? Isn't the action always bid?

magnusfalt replied:
Sure i can remove the action, bidbox_.. and just use bid and still check the rspond_to?

crericha replied:
Either `respond_to?` or something like a `can_bid?(player, entity)` method that is always false in the base class.

tobymao replied:
what is all of this?

crericha commented on assets/app/view/game/player.rb:
```diff
@@ -101,6 +101,13 @@ def render_info
+          trs.concat([
+             h(:tr, [
+               h(:td, 'Bidding tokens'),
+               h('td.right', "#{@game.active_step.bidding_tokens(@player)} / #{@game.bidding_token_per_player}"),
```
Using round state would be better than active_step for managing bidding tokens.

crericha replied:
Still would like to see `@round.bidding_tokens` using the `round_state` method. Ping me if you aren't sure what I mean.

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -350,6 +356,68 @@ def render_buy_input_interval(company)
           ])]
         end
+
+        def render_bidbox(companies)
```
I wonder if a special render is needed. Other games include privates in the stock round (1860 and 1824), but for purchasing. It would reduce complexity if we did it the same way here and only added the bid button logic. I suggest taking a look.

magnusfalt replied:
I check how all other have done it, and also talked with Per who implements 1824. I want to seperate the bidboxes in its own row.

magnusfalt replied:
And per also wanted to seperate the minor bids from the rest of the privates.

tobymao replied:
if you can do it the generic way, please remove all of this code

crericha commented on assets/app/view/game/company.rb:
```diff
@@ -57,6 +57,8 @@ def render
             fontSize: '90%',
           }
+          header_style[:background] = @company.header_color if @company.respond_to?(:header_color)
+          header_style[:color] = @company.header_text_color if @company.respond_to?(:header_text_color)
```
Would it make sense to add header_color and header_text_color to company and default to the standard colors? That way we don't have to add any additional respond_to? nonsense, simplifying the code.

magnusfalt replied:
Since im new here i dont know how much i can add to the "base" code. Its no problem to add it to the base company.

crericha replied:
Its definitely a judgement call and should be done carefully as it affects all games. My general rule of thumb is if the functionality would be useful to other games, I add it to the base class. If its only useful to my game, I add it to my specific game classes. In this case, I could see other games taking advantage of it. I'll probably use this for 1828 if it becomes available in the code base.

tobymao replied:
remove this stuff

magnusfalt replied:
Remove all the header_ code, so the minors and consession is the yellow color you mean @tobymao ?

tobymao replied:
either make it all generic, or remove it, if you want to keep the PR simple, i'd remove all of this specialized stuff for now

crericha replied:
@tobymao, 1822 needs a way for the player to know who can use privates - majors only, players only, majors and minors. The print version uses the header color to distinguish. Seems like an unobtrusive way to distinguish here too. Just my two cents.

roseundy replied:
Toby has beaten into my head the necessity to keep UI code out of the engine code. Putting a header_color method into company violates that. Better to have a method that indicates who may use a company and have the UI code look at that to decide on header colors.

magnusfalt replied:
Im all for keeping ui code from the engine. But we have color on the major companies, just wanted it on the minors, concession to differ them during the bid. But i will remove it.

tobymao replied:
if you need the color, then allow every game to have color and have it be a part of the config, don't have respond_to etc 

happy to chat more on slack if you want to discuss

crericha commented on lib/engine/g_1822/company.rb:
```diff
@@ -0,0 +1,24 @@
+        @header_text_color = opts[:header_text_color] || :black
+        @removed = false
+
+        super(**opts)
```
Recommend calling super first, so that your instance vars don't accidently get overridden.

crericha replied:
I know that's not the case now, but if someone adds one of those types to the base class, it will be an issue.

magnusfalt replied:
Sure i can do that. But i have looked and super is really all over the place in the code.

crericha commented on lib/engine/g_1822/company.rb:
```diff
@@ -0,0 +1,24 @@
+        @header_name = opts[:header_name] || 'PRIVATE COMPANY'
+        @header_color = game.registered_color(opts[:header_color]) || :yellow
+        @header_text_color = opts[:header_text_color] || :black
+        @removed = false
```
Do you need to track removed or simply remove in the game class when the action happens? `@companies.delete(company)`

magnusfalt replied:
It was a beginning of a thought. I do remove it, therefor i can remove the remove attribute.

tobymao replied:
remove this class

magnusfalt replied:
Hmm i still need the name for the company. If we gonna remove the coloring name is still needed 'PRIVATE COMPANY' in the view.

magnusfalt replied:
Ok will remove it.

crericha commented on lib/engine/g_1822/minor.rb:
```diff
@@ -0,0 +1,25 @@
+  module G1822
+    class Minor < Minor
+      attr_accessor :type, :removed, :share_price, :max_share_price
+      attr_writer :par_price
```
1867 has minors with share prices. Is there a way we can standardize instead of duplicating?

magnusfalt replied:
Same as color, dont know much i can add to the base code.

crericha replied:
Code duplication is bad and should be avoided where possible.

tobymao replied:
remove this class

magnusfalt replied:
Do you mean move the share_price, par_price and max_share_price to base class?

tobymao replied:
you don't need those things in a minor. if you have shares, then it's a corporation not a minor, thus you should remove this 1822 specific minor

crericha commented on lib/engine/game/g_1822.rb:
```diff
@@ -58,12 +78,68 @@ def operating_round(round_num)
+        # Randomize from preset seed to get same order
+        @companies.sort_by! { rand }
+
+        p1 = @companies.find { |c| !c.nil? && c.id == :P1 }
```
I think you want 'P1' here. Symbols == Strings in JS land, but not so in Ruby once you hit the server. (Note: Hotseat games are all in JS, so that's why you wouldn't see it yet.)

magnusfalt replied:
Ahh ok, will change it.

crericha commented on lib/engine/game/g_1822.rb:
```diff
@@ -58,12 +78,68 @@ def operating_round(round_num)
+        @companies.delete(p1)
+        @companies.unshift(p1)
+
+        c1 = @companies.find { |c| !c.nil? && c.id == :C1 }
```
'C1'

crericha commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,76 @@
+          super
+        end
+
+        def setup
```
Don't need to override if only calling super.

magnusfalt replied:
Im miles away before im done with that file, just added because its gonna med populated with code soon.

crericha commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,76 @@
+          company = bid.company
+          price = bid.price
+
+          return unless (minor = @game.minors.find { |m| m.id == company.id })
```
You've detected a broken game here. Raise an error to reject the last action rather than silently continue on.

crericha replied:
In general, raise an error when something happens that shouldn't be possible.

magnusfalt replied:
True will fix, shouldnt need to do the return.

crericha commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,76 @@
+          @game.bank.spend(100, minor)
+
+          hex = @game.hex_by_id(minor.coordinates)
+          hex.tile.cities[minor.city || 0].place_token(minor, minor.next_token)
```
If I recall correctly, token placement is part of first turn housekeeping. It shouldn't block another minor that runs earlier from running through the city.

magnusfalt replied:
Thats true, but as you see in the rest of the method par price and other stuff is hardcoded. It is not finished, just places the token in that part of the code just not to break the game.

crericha replied:
I add TODO comments for those types of things, so I don't forget. That's just a personal preference though.

magnusfalt replied:
Im not used to add TODO that much. But since its hard for the reviews to know how i think i might add a bit more todo's code that is not finished. But in this state most of the code isnt finished :)

crericha replied:
Sure. If you don't like the TODOs, including comments in the PR can provide context to reviewers too. Whatever works best for you and your workflow. :-)

crericha commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,76 @@
+          share_price = @game.stock_market.par_prices.find { |pp| pp.price == 50 }
+          @game.stock_market.set_par(minor, share_price)
+
+          @game.companies.delete(company)
```
Why is the company being deleted? Its still in the game.

magnusfalt replied:
No the company for the minor is a placeholder to get the bidding to work without to rework the existing bidding mechanics. When the minor floats the company for the minor is removed. In the real game you just put the minors in its biddingboxes.

crericha commented on lib/engine/step/g_1822/bidbox_auction.rb:
```diff
@@ -0,0 +1,65 @@
+        end
+
+        def remove_from_auction(entity)
+          super
```
Don't override methods that you only call super in. There are more below.

crericha commented on lib/engine/step/g_1822/bidbox_auction.rb:
```diff
@@ -0,0 +1,65 @@
+
+        def bids_for_player(player, only_highest_bid_committed = false)
+          @bids.values.map do |bids|
+            if only_highest_bid_committed
```
Since this is in 1822 code, maybe rename the argument to `only_committed_bids`. Also, prefer keyword arguments (`only_highest_bid_committed: false`) over defaulting arguments, so that order doesn't matter.

magnusfalt replied:
Sure will change it.

crericha commented on lib/engine/step/g_1822/buy_sell_par_shares.rb:
```diff
@@ -0,0 +1,79 @@
+          actions << 'bidbox_minors'
+          actions << 'bidbox_concessions'
+          actions << 'bidbox_privates'
+          actions << 'bidding_tokens'
```
Yeah, the extra actions feel unnecessary. Maybe I'm missing something.

crericha replied:
Also, don't include if the player can't lay or increase a bid due to available cash. This is needed to auto-skip that player if they have no actions.

magnusfalt replied:
As commented above, i can change just to have the bid left.

magnusfalt replied:
Yes good ide. There are a few more places where autopass can be added.

crericha commented on lib/engine/step/g_1822/buy_train.rb:
```diff
@@ -0,0 +1,20 @@
+          return [] if entity != current_entity
+
+          return %w[sell_shares buy_train] if president_may_contribute?(entity)
+          return %w[buy_train pass] if can_buy_train?(entity)
```
Is this for minors? Does 1867 do the same?

magnusfalt replied:
Was a long time since i played 1867 so i dont remember. The code is not done, i just took the base function and removed the
`return [] if entity.minor?`
so the game doesnt crash.

crericha replied:
The good news is you may be able to use a lot of Jen's 1867 code for minors. I haven't looked into how she implemented it, so don't know for sure.

jenf replied:
1867 minors aren't actually minors but real corporations with a type set to :minor, :major's are major's

jenf replied:
1867 minors needed share price and not going down this approach was going to be quite painful.

crericha replied:
@magnusfalt, I recommend going the route @jenf did. It sounds like it'll make for an easier implementation. The Minors class is intended for non-share minors like in 1846.

tobymao commented on assets/app/view/game/company.rb:
```diff
@@ -97,8 +99,9 @@ def render
+          header_name = (@company.respond_to?(:header_name) ? @company.header_name : 'PRIVATE COMPANY')
           children = [
-            h(:div, { style: header_style }, 'PRIVATE COMPANY'),
+            h(:div, { style: header_style }, header_name),
```
remove this

magnusfalt commented:
The changes in the last commit:

* Removed the derived class Company, move the color and text_color to the base Company class for all other to enjoy.
* With the removal of Company the type was also removed, therefore in places where i need to sperate the 3 different companytypes i look at the first char in the id (sym) property.
* Removed the derived class Minor, this was needed to get a minor to have a stock value in the stockmarket. Instead the minor will float as a corporation with the type minor.
* Removed the bidbox logic from the stock view. Im using the render_bank_companies instead with a small addition of the 'bid' action and the method render_bid_input.
* The method render_bid_input now uses the Bid component.
* Changed the Bid component to also accept a company, this is done to get the correct property of the return array which calls the process_bid method.

jenf commented on lib/engine/game/g_1822.rb:
```diff
@@ -33,19 +33,35 @@ class G1822 < Base
+        stock_round
+      end
+
+      def operating_order
```
maybe we should separate this and sorted_corporations out into an include since it's the same for 1867 and 1822

magnusfalt replied:
That would be a good idea, i found more code that we can share in a include. i e with 1817 i have copied:

`      def format_currency(val)
        # On dividends per share can be a float
        # But don't show decimal points on all
        return format('$%.1<val>f', val: val) if val.is_a?(Float) && (val % 1 != 0)

        self.class::CURRENCY_FORMAT_STR % val
      end`

I can wait to extract them into an include until we see how much more i need from any specific title that isnt already in an include.

jenf commented on lib/engine/step/g_1822/buy_sell_par_shares.rb:
```diff
@@ -0,0 +1,76 @@
+        end
+
+        def description
+          'Bid, convert consession or buy/sell shares'
```
concession

crericha commented on assets/app/view/game/bid.rb:
```diff
@@ -8,14 +8,16 @@ module Round
         needs :entity
-        needs :corporation
+        needs :corporation, default: nil
+        needs :company, default: nil
```
Generalize corporation and company into one type.

magnusfalt replied:
Sure, im have mostly programmed in strong type languages. Not used to use the same variable for different objects.

crericha commented on assets/app/view/game/bid.rb:
```diff
@@ -29,6 +31,7 @@ def render
             process_action(Engine::Action::Bid.new(
               @entity,
               corporation: @corporation,
+              company: @company,
```
Again, generalize here. Under-the-hood, you can use `corporation?` and `company?` to store the type for serialization purposes.

magnusfalt replied:
Do we have a good name for the combination of corporation and company the same variable. Entity is already used for the player.

crericha replied:
Perhaps change entity to bidder and then use entity?

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -350,6 +354,16 @@ def render_buy_input_interval(company)
+        def render_bid_input(company)
+          return [] if @step.max_bid(@current_entity, company) < @step.min_bid(company)
+          return [] if @current_actions.include?('bidding_tokens') &&
+            @step.highest_player_bid?(@current_entity, company)
```
Is this true, you can't increase your bid if winning? I understand why you wouldn't want to, but is it against the rules?

magnusfalt replied:
Yes it is

magnusfalt replied:
Yes you cant move the cude if you are the highest:

4.9.1 Move a previously placed bidding token to a higher-valued empty space on the same item. This is only allowed if
another player has placed a higher bid.

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -350,6 +354,16 @@ def render_buy_input_interval(company)
+          return [] if @current_actions.include?('bidding_tokens') &&
+            @step.highest_player_bid?(@current_entity, company)
+          return [] if @current_actions.include?('bidding_tokens') && @step.find_bid(@current_entity, company).nil? &&
+            @step.bidding_tokens(@current_entity).zero?
```
The UI shouldn't have to make these logic decisions. Can we replace with `@step.can_bid?(@current_entity, company)`?

magnusfalt replied:
Yep that i can do.

crericha commented on lib/engine/game/base.rb:
```diff
@@ -203,6 +203,8 @@ class Base
 
       CORPORATIONS = [].freeze
 
+      MINORS = [].freeze
```
Is this still needed?

magnusfalt replied:
I was getting an warning that is wasnt declared, but i can revert it.

crericha commented on lib/engine/step/g_1822/bidbox_auction.rb:
```diff
@@ -0,0 +1,57 @@
+          yield company, bids if bids.size > 1
+        end
+
+        def bids_for_player(player, only_committed_bids = false)
```
Prefer keyword arguments for optional arugments - `only_committed_bids: false`

magnusfalt commented:
Changed the things pointed out by Jennifer and Chris.

crericha commented on assets/app/view/game/buy_trains.rb:
```diff
@@ -103,9 +103,11 @@ def render
           store(:active_shell, nil, skip: true)
         end
+        puts 'Render buy trains 5'
 
```
Remove debug in this file

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -357,12 +357,9 @@ def render_buy_input_interval(company)
-            @step.highest_player_bid?(@current_entity, company)
-          return [] if @current_actions.include?('bidding_tokens') && @step.find_bid(@current_entity, company).nil? &&
-            @step.bidding_tokens(@current_entity).zero?
+          return [] if @current_actions.include?('bidding_tokens') && !@step.can_bid?(@current_entity, company)
```
Why do we need the bidding_tokens action? Can't we use bid? Isn't the player card the only place in the UI that needs to know about bidding_tokens?

crericha replied:
Refactor line 359 to use `can_bid?` and 1822 bidding is common with other game bidding within the UI hooray.

magnusfalt replied:
The can_bid? method isnt in the normal Auctioner module which uses the 'bid' action. Any game that uses bid in the stockround will then throw an error when trying to access can_bid?.  bidding_tokens is also used in the player.rb to see how many tokens you have left to bid. Dont know how to show them otherwise?

magnusfalt replied:
if we cant use @step.respond_to?(:can_bid?) then if we gonna use bid, we need to have a dummy method in Auctioner with can_bid? not to break for every other game.

crericha replied:
What I meant by refactor is move line 359 to be a method called `can_bid?` within Auctioner. Keep the logic and game differentiation to the engine code as much as possible. Working on the UI can often include a bunch of refactoring to more generalize the code to include your use cases.

See my other comment on the player card.

magnusfalt replied:
I can remove bidding token on the player and print a log message if this is a better solution.

magnusfalt replied:
Sure you can count your bids, as well. Just wanted the user to see in one place. Committed cash, Available cash and how many tokens he/she have left.

magnusfalt replied:
Ok i see i have made lots of comments without seeing your new ones :)

crericha replied:
Definitely keep the bidding tokens on the player card. Its important info that they shouldn't have to calculate for each player.

crericha commented on assets/app/view/game/player.rb:
```diff
@@ -101,6 +101,13 @@ def render_info
+               h(:td, 'Bidding tokens'),
+               h('td.right', "#{@game.active_step.bidding_tokens(@player)} / #{@game.bidding_token_per_player}"),
+             ]),
+           ]) if @game.active_step&.current_actions&.include?('bidding_tokens')
```
This is the only other place using the bidding_tokens action. Change to `@game.active_step.respond_to?(:bidding_tokens)`, so we don't need a special action.

crericha commented on assets/app/view/game/round/stock.rb:
```diff
@@ -296,15 +296,19 @@ def render_bank_companies
         end
 
         def render_buy_input(company)
-          return [] unless @current_actions.include?('buy_company')
```
Why did you remove this line?

magnusfalt replied:
I moved it to row 303 to be more cohesive with the bid action i added.

```ruby
inputs.concat(render_buy_input(company)) if @current_actions.include?('buy_company')
inputs.concat(render_bid_input(company)) if @current_actions.include?('bid')
```

crericha commented on lib/engine/step/auctioner.rb:
```diff
@@ -77,6 +77,10 @@ def max_bid(_entity, _company)
       end
 
+      def can_bid?(_entity, _company)
+        true
```
Add `return false if max_bid(entity, company) < min_bid(company)` to maintain the old logic.

Also, make sure to test at least one other game to verify your changes didn't break the bidding UI.

magnusfalt replied:
Hmm there was not old logic its a new function. But yes it makes sense to have that locig there.

crericha replied:
The logic was moved out of the UI in the refactor, but it was not placed here. That's what I'm referring too. We still need to maintain the same behavior.

magnusfalt replied:
I really dont follow, the render_bid_input method is new to this PR. Therefore the logic in it can not be old.

magnusfalt replied:
But i agree the default behaviour should be the logic of `max_bid(entity, company) < min_bid(company)`

crericha replied:
My apologies. I confused the file with ` assets/app/view/game/bid.rb`. Thanks for clarifying. 

Since 1822 is the only current game to access bid from the stock round, you can either remove `can_bid?` from the base auctioner or add in the basic logic outlined above. Whichever way you prefer.

magnusfalt replied:
There are more games to use bid in the stockround, just not the way i do it. In method `render_pre_ipo` there is a line 

`children << h(Bid, entity: @current_entity, corporation: corporation) if @current_actions.include?('bid')`

Im guessing its from 1817.

magnusfalt replied:
Since in the base auctioneer the methods `min_bid` and `max_bid` do a `raise NotImplementedError`. Shall i do the same for `can_bid?`

tobymao commented on lib/engine/game/g_1822.rb:
```diff
@@ -62,13 +78,94 @@ def operating_round(round_num)
+      end
+
+      def bidbox_minors
+        @companies.select { |c| c.id[0] == 'M' && (c.owner.nil? || c.owner == @bank) }.map.first(4)
```
what's this map here for?

don't call owner.nil?

&& !c.owner || c.owner == @bank

magnusfalt replied:
Just a lack of knowledge of ruby, i probably saw code something like it and did it about the same way.

tobymao commented on lib/engine/game/g_1822.rb:
```diff
@@ -62,13 +78,94 @@ def operating_round(round_num)
+
+      def setup_bidboxes
+        # Set the owner to bank for the companies up for auction this stockround
+        bidbox_minors.map do |minor_company|
```
call this a minor

tobymao commented on lib/engine/game/g_1822.rb:
```diff
@@ -62,13 +78,94 @@ def operating_round(round_num)
+          concessions.owner = @bank
+        end
+
+        bidbox_privates.map do |private_company|
```
call this a company

tobymao commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,109 @@
+        def finish_round
+          @game.bidbox_minors.map do |minor_company|
+            bid = highest_bid(minor_company)
+            if bid.nil?
```
if (bid = highest_bid...)
else

tobymao replied:
why are you using map? you should be using each

tobymao commented on lib/engine/round/g_1822/stock.rb:
```diff
@@ -0,0 +1,109 @@
+          end
+
+          @game.bidbox_concessions.map do |concessions|
+            bid = highest_bid(concessions)
```
same

tobymao commented on lib/engine/step/g_1822/bidbox_auction.rb:
```diff
@@ -0,0 +1,57 @@
+        end
+
+        def highest_player_bid?(player, company)
+          return false if find_bid(player, company).nil?
```
return false unless ...

tobymao commented on lib/engine/step/g_1822/buy_sell_par_shares.rb:
```diff
@@ -0,0 +1,82 @@
+          super
+
+          @bid_actions = 0
+          @bidders = (@round.bidders.nil? ? Hash.new { |h, k| h[k] = [] } : @round.bidders)
```
no nil

tobymao commented on lib/engine/step/g_1822/buy_sell_par_shares.rb:
```diff
@@ -0,0 +1,82 @@
+
+          @bid_actions = 0
+          @bidders = (@round.bidders.nil? ? Hash.new { |h, k| h[k] = [] } : @round.bidders)
+          @bids = @round.bids unless @round.bids.nil?
```
same

tobymao commented on lib/engine/step/g_1822/buy_sell_par_shares.rb:
```diff
@@ -0,0 +1,82 @@
+        def can_bid?(entity, company)
+          return false if max_bid(entity, company) < min_bid(company) || highest_player_bid?(entity, company)
+
+          !(find_bid(entity, company).nil? && bidding_tokens(entity).zero?)
```
nil

magnusfalt commented:
Improved the code with Tobys suggestions

tobymao commented on lib/engine/game/g_1822.rb:
```diff
@@ -62,13 +78,94 @@ def operating_round(round_num)
+
+      def setup_bidboxes
+        # Set the owner to bank for the companies up for auction this stockround
+        bidbox_minors.map do |minor|
```
you didn't fix these

tobymao commented on lib/engine/game/g_1822.rb:
```diff
@@ -62,13 +78,94 @@ def operating_round(round_num)
+          concessions.owner = @bank
+        end
+
+        bidbox_privates.map do |private|
```
don't call it a private it's a reserved keyword
